internal/pkg/docker/scheduler/scheduler.go

```diff
@@ -0,0 +1,312 @@
+	//   returned by it are reported by the scheduler.
+	// - if Stop() is called _before_ the scheduler picks up this
+	//   action, then this action is skipped.
+	select {
+	case <-s.stopped:
+	case s.actions <- &runTaskAction{
+		task: task,
+	}:
+	}
+}
+
+type runTaskAction struct {
+	task Task
+}
+
+func (r *runTaskAction) Do(s *Scheduler) error {
+	// we no longer care about errors from the old task
+	taskID := s.curTaskID.Add(1)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// cancelCtxOnStop calls cancel if Stop() is called before ctx finishes.
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		select {
+		case <-ctx.Done():
+		case <-s.stopped:
+			cancel()
+		}
+	}()
+
+	if taskID == 1 {
+		// start the pause container
+		opts := s.pauseRunOptions(r.task)
+		s.run(-1, opts)
+		if err := s.waitForContainerToStart(ctx, opts.ContainerName); err != nil {
+			return fmt.Errorf("wait for pause container to start: %w", err)
+		}
+	} else {
+		// ensure no pause container changes
+		curOpts := s.pauseRunOptions(s.curTask)
+		newOpts := s.pauseRunOptions(r.task)
+		if !maps.Equal(curOpts.EnvVars, newOpts.EnvVars) ||
+			!maps.Equal(curOpts.Secrets, newOpts.Secrets) ||
+			!maps.Equal(curOpts.ContainerPorts, newOpts.ContainerPorts) {
+			return errors.New("new task requires recreating pause container")
+		}
+
+		if err := s.stopTask(ctx, s.curTask); err != nil {
+			return fmt.Errorf("stop existing task: %w", err)
+		}
+	}
+
+	s.curTask = r.task
+	for name, ctr := range r.task.Containers {
+		name, ctr := name, ctr
+		s.run(taskID, s.containerRunOptions(name, ctr))
+	}
+
+	return nil
+}
+
+// Stop stops the current running task containers and the Scheduler. Stop is
+// idempotent and safe to call multiple times. Calls to RunTask() after calling Stop
+// do nothing.
+func (s *Scheduler) Stop() {
+	s.stopOnce.Do(func() {
+		close(s.stopped)
+		s.actions <- &stopAction{}
+	})
+}
+
+type stopAction struct{}
+
+func (a *stopAction) Do(s *Scheduler) error {
+	defer s.wg.Done()  // for the scheduler
+	s.curTaskID.Add(1) // ignore runtime errors
+
+	// collect errors since we want to try to clean up everything we can
+	var errs []error
+	if err := s.stopTask(context.Background(), s.curTask); err != nil {
+		errs = append(errs, err)
+	}
+
+	// stop pause container
+	if err := s.docker.Stop(context.Background(), s.containerID("pause")); err != nil {
+		errs = append(errs, fmt.Errorf("stop %q: %w", "pause", err))
+	}
+
+	return errors.Join(errs...)
+}
+
+// stopTask calls `docker stop` for all containers defined by task.
+func (s *Scheduler) stopTask(ctx context.Context, task Task) error {
+	if len(task.Containers) == 0 {
+		return nil
+	}
+
+	errCh := make(chan error)
+	for name := range task.Containers {
+		name := name
+		go func() {
+			if err := s.docker.Stop(ctx, s.containerID(name)); err != nil {
+				errCh <- fmt.Errorf("stop %q: %w", name, err)
+				return
+			}
+			errCh <- nil
```

Comment:
Is this sending a `nil` to `errCh` so that we can use ⬇️  to determine whether all `docker stop` have finished?
```go
if len(errs) == len(task.Containers) {
	break
}
```

Reply:
This is what I gathered yeah 👍

Reply:
yeah, you're exactly right! i'm looking for one response per container. i'll add a comment to clarify this!